Cloud Build Failure Reporter: integration-test-nightly (langchain-alloydb-testing) failed

Cloud Build Failure Reporter found test failure for [**integration-test-nightly (langchain-alloydb-testing)** ](https://github.com/googleapis/langchain-google-alloydb-pg-python/runs/34412640973) at [8ad96221238c93695730514c30613cb72d8f7959](https://github.com/googleapis/langchain-google-alloydb-pg-python/commit/8ad96221238c93695730514c30613cb72d8f7959). Please fix the error and then close the issue after the **integration-test-nightly (langchain-alloydb-testing)** test passes.

```
        async with self._engine._pool.connect() as conn:
            await conn.execute(text(create_extension_query))
            await conn.commit()
        extension_version_query = "SELECT extversion FROM pg_extension WHERE extname = 'google_ml_integration';"
        result = await self.__query_db(extension_version_query)
        version = result[0]["extversion"]
>       return float(version)
E       ValueError: could not convert string to float: '1.4.2'
```

--- a/src/langchain_google_alloydb_pg/model_manager.py
+++ b/src/langchain_google_alloydb_pg/model_manager.py
@@ -90,7 +90,9 @@ class AlloyDBModelManager:
             :class: `AlloyDBModel` object of the specified model if it exists otherwise `None`.
 
         """
-        result = await self._engine._run_as_async(self.__aget_model(model_id=model_id))
+        result = await self._engine._run_as_async(
+            self.__aget_model(model_id=model_id)
+        )
         return result
 
     async def alist_models(self) -> list[AlloyDBModel]:
@@ -133,7 +135,11 @@ class AlloyDBModelManager:
         """
         await self._engine._run_as_async(
             self.__acreate_model(
-                model_id, model_provider, model_type, model_qualified_name, **kwargs
+                model_id,
+                model_provider,
+                model_type,
+                model_qualified_name,
+                **kwargs,
             )
         )
 
@@ -157,7 +163,7 @@ class AlloyDBModelManager:
         """
         extension_version = await self.__fetch_google_ml_extension()
         db_flag = await self.__fetch_db_flag()
-        if extension_version < 1.3:
+        if extension_version < "1.3":
             raise Exception(
                 "Please upgrade google_ml_integration EXTENSION to version 1.3 or above."
             )
@@ -265,7 +271,7 @@ class AlloyDBModelManager:
             await conn.execute(text(query))
             await conn.commit()
 
-    async def __fetch_google_ml_extension(self) -> float:
+    async def __fetch_google_ml_extension(self) -> str:
         """Creates the Google ML Extension if it does not exist and returns the version number (Default creates version 1.3)."""
         create_extension_query = """
         DO $$
@@ -283,7 +289,7 @@ class AlloyDBModelManager:
         extension_version_query = "SELECT extversion FROM pg_extension WHERE extname = 'google_ml_integration';"
         result = await self.__query_db(extension_version_query)
         version = result[0]["extversion"]
-        return float(version)
+        return version
 
     async def __fetch_db_flag(self) -> str:
         """Fetches the enable_model_support DB flag."""

--- a/src/langchain_google_alloydb_pg/model_manager.py
+++ b/src/langchain_google_alloydb_pg/model_manager.py
@@ -15,8 +15,7 @@
 # TODO: Remove below import when minimum supported Python version is 3.10
 from __future__ import annotations
 
-from dataclasses import dataclass
-from typing import Optional, Sequence
+from typing import Any, Optional, Sequence
 
 from sqlalchemy import text
 from sqlalchemy.engine.row import RowMapping
@@ -37,7 +36,7 @@ class AlloyDBModel:
         input_transform_fn: Optional[str],
         output_transform_fn: Optional[str],
         generate_headers_fn: Optional[str] = None,
-        **kwargs,
+        **kwargs: Any,
     ):
         self.model_id = model_id
         self.model_request_url = model_request_url
@@ -48,6 +47,7 @@ class AlloyDBModel:
         self.model_auth_id = model_auth_id
         self.input_transform_fn = input_transform_fn
         self.output_transform_fn = output_transform_fn
+        # List models is returning column name "header_gen_fn"
         self.generate_headers_fn = generate_headers_fn or kwargs.get("header_gen_fn")
 
 

--- a/src/langchain_google_alloydb_pg/model_manager.py
+++ b/src/langchain_google_alloydb_pg/model_manager.py
@@ -192,16 +192,7 @@ class AlloyDBModelManager:
 
         """
         query = f"""SELECT * FROM
-                google_ml.list_model('{model_id}')
-                AS t(model_id VARCHAR,
-                model_request_url VARCHAR,
-                model_provider google_ml.model_provider,
-                model_type google_ml.model_type,
-                model_qualified_name VARCHAR,
-                model_auth_type google_ml.auth_type,
-                model_auth_id VARCHAR,
-                input_transform_fn VARCHAR,
-                output_transform_fn VARCHAR)"""
+                google_ml.list_model('{model_id}')"""
 
         try:
             result = await self.__query_db(query)

--- a/src/langchain_google_alloydb_pg/model_manager.py
+++ b/src/langchain_google_alloydb_pg/model_manager.py
@@ -90,9 +90,7 @@ class AlloyDBModelManager:
             :class: `AlloyDBModel` object of the specified model if it exists otherwise `None`.
 
         """
-        result = await self._engine._run_as_async(
-            self.__aget_model(model_id=model_id)
-        )
+        result = await self._engine._run_as_async(self.__aget_model(model_id=model_id))
         return result
 
     async def alist_models(self) -> list[AlloyDBModel]:

--- a/src/langchain_google_alloydb_pg/model_manager.py
+++ b/src/langchain_google_alloydb_pg/model_manager.py
@@ -24,18 +24,31 @@ from sqlalchemy.engine.row import RowMapping
 from .engine import AlloyDBEngine
 
 
-@dataclass
 class AlloyDBModel:
-    model_id: str
-    model_request_url: Optional[str]
-    model_provider: str
-    model_type: str
-    model_qualified_name: str
-    model_auth_type: Optional[str]
-    model_auth_id: Optional[str]
-    header_gen_fn: Optional[str]
-    input_transform_fn: Optional[str]
-    output_transform_fn: Optional[str]
+    def __init__(
+        self,
+        model_id: str,
+        model_request_url: Optional[str],
+        model_provider: str,
+        model_type: str,
+        model_qualified_name: str,
+        model_auth_type: Optional[str],
+        model_auth_id: Optional[str],
+        input_transform_fn: Optional[str],
+        output_transform_fn: Optional[str],
+        generate_headers_fn: Optional[str] = None,
+        **kwargs,
+    ):
+        self.model_id = model_id
+        self.model_request_url = model_request_url
+        self.model_provider = model_provider
+        self.model_type = model_type
+        self.model_qualified_name = model_qualified_name
+        self.model_auth_type = model_auth_type
+        self.model_auth_id = model_auth_id
+        self.input_transform_fn = input_transform_fn
+        self.output_transform_fn = output_transform_fn
+        self.generate_headers_fn = generate_headers_fn or kwargs.get("header_gen_fn")
 
 
 class AlloyDBModelManager:
@@ -201,7 +214,7 @@ class AlloyDBModelManager:
                 model_qualified_name VARCHAR,
                 model_auth_type google_ml.auth_type,
                 model_auth_id VARCHAR,
-                header_gen_fn VARCHAR,
+                generate_headers_fn VARCHAR,
                 input_transform_fn VARCHAR,
                 output_transform_fn VARCHAR)"""
 

--- a/src/langchain_google_alloydb_pg/model_manager.py
+++ b/src/langchain_google_alloydb_pg/model_manager.py
@@ -33,6 +33,7 @@ class AlloyDBModel:
     model_qualified_name: str
     model_auth_type: Optional[str]
     model_auth_id: Optional[str]
+    header_gen_fn: Optional[str]
     input_transform_fn: Optional[str]
     output_transform_fn: Optional[str]
 
@@ -192,7 +193,17 @@ class AlloyDBModelManager:
 
         """
         query = f"""SELECT * FROM
-                google_ml.list_model('{model_id}')"""
+                google_ml.list_model('{model_id}')
+                AS t(model_id VARCHAR,
+                model_request_url VARCHAR,
+                model_provider google_ml.model_provider,
+                model_type google_ml.model_type,
+                model_qualified_name VARCHAR,
+                model_auth_type google_ml.auth_type,
+                model_auth_id VARCHAR,
+                header_gen_fn VARCHAR,
+                input_transform_fn VARCHAR,
+                output_transform_fn VARCHAR)"""
 
         try:
             result = await self.__query_db(query)

--- a/tests/test_model_manager.py
+++ b/tests/test_model_manager.py
@@ -64,6 +64,7 @@ class TestAlloyDBModelManager:
         )
         yield engine
         await engine.close()
+        await engine._connector.close()
 
     @pytest_asyncio.fixture(scope="module")
     async def model_manager(self, engine):

